predict.py

```diff
@@ -1,94 +1,280 @@
+            video_length = new_video_length
 
-        generator = torch.Generator("cuda").manual_seed(seed)
-
-        output = self.pipe(
-            prompt=prompt,
-            height=height,
-            width=width,
-            num_frames=num_frames,
-            num_inference_steps=num_inference_steps,
-            generator=generator,
-            guidance_scale=guidance_scale,
-        ).frames[0]
-
-        output_path = "/tmp/output.mp4"
-        export_to_video(output, output_path, fps=fps)
-        return Path(output_path)
+        # ensure that height/width/video length result in a sequence length that is divisible by WORLD_SIZE
+        seq_length = self._get_seq_length(height, width, video_length)
+        if seq_length % self.WORLD_SIZE != 0:
+            new_height = self._adjust_height(height, width, video_length, self.WORLD_SIZE)
+            print(f"Warning: sequence length {seq_length} is not divisible by WORLD_SIZE {self.WORLD_SIZE}, rounding height up to {new_height}")
+            height = new_height
+
+        # send request to the server and wait
+        save_path = f"/tmp/output_{seed}.mp4"
+        predict_args = {
+            "prompt": prompt,
+            "height": height,
+            "width": width,
+            "video_length": video_length,
+            "infer_steps": infer_steps,
+            "embedded_guidance_scale": embedded_guidance_scale,
+            "seed": seed,
+            "fps": fps,
+            "output_path": save_path
+        }
+
+        self._request_and_wait(predict_args)
+
+        return Path(save_path)
+
+
+    def _get_signal_handler(self):
+        
+        def signal_handler(signum, frame):
+            print(f"Received signal: {signum}")
+
+            # pass on cancellation message to the server
+            # this will result in the server's worker processes stopping
+            # and returning once the current diffusion step has finished
+            cancel_url = f"http://localhost:{self.HTTP_PORT}/cancel"
+            response = requests.post(cancel_url)
+            print(f"Cancel response: {response.json()}")
+            
+            # set internal flag to true, this causes us to stop polling the server
+            self.request_cancelled = True
+            assert response.status_code == 200
+        
+        return signal_handler
+    
+
+    def _request_and_wait(self, predict_args, timeout=20):
+        """
+        sent a prediction request to the server and wait, wait for the server to finish generating the video
+        once the server has finished generating the video, it will return a 200 response and the video will be
+        saved to predict_args['output_path']
+        """
+        # make sure server is running, if not restart it
+        if not self.server_process.is_alive():
+            self._restart_server()
+        
+        # make sure server is ready to accept requests
+        # if it is not up for 20 seconds, restart it
```

Comment:
nit - this not up for 20 seconds isn't entirely true, depends on the timeout. might want to update the comment.